typing/typecore.ml

```diff
@@ -2948,59 +2947,107 @@ let rec approx_type env sty =
         let tyl = List.map (approx_type env) ctl in
         newconstr path tyl
       end
-  | Ptyp_poly (_, sty) ->
+  | Ptyp_poly _ ->
       approx_type env sty
   | _ -> newvar ()
 
-let rec type_approx env sexp =
-  match sexp.pexp_desc with
-    Pexp_let (_, _, e) -> type_approx env e
-  | Pexp_fun (p, _, spat, e) ->
-      let marg =
-        if has_local_attr_pat spat then Alloc_mode.local
-        else Alloc_mode.newvar ()
+let type_pattern_approx env spat ty_expected =
+  match spat.ppat_desc with
+  | Ppat_constraint(_, ({ptyp_desc=Ptyp_poly _} as sty)) ->
```

Comment:
You're not missing anything, you just forgot to turn on principal mode. With `-principal` you get:
```ocaml
4 | let rec f K2 = f K1 + f K2;;
                     ^^
Warning 18 [not-principal]: this type-based constructor disambiguation is not principal.
```

The polymorphic constraint case is more important than the others because if we don't handle it then `type_approax` will guess a monomorphic type for the parameter and it won't unify with the actual type.